File: packages/vault/src/bootstrap.ts
// Vault bootstrap + enrollment: the owner-administrative acts before the
// gateway can authenticate anyone (first device = chicken-and-egg).

import { randomBytes } from "node:crypto";

import type { VaultDb } from "./db.js";
import type { FilterClause, Risk } from "./gateway/types.js";
import { setDeviceTrust } from "./grant/device-trust.js";
import { nowIso, uuidv7 } from "./ids.js";

export interface BootstrapResult {
  vaultId: string;
  displayName: string;
  /**
   * The vault's SELF party (#916, ruling ONT-05): the person as DATA. It
   * confers no permission — authority is gateway-side (`vault_owners`) and in
   * `share_authority` — and the column it lands in is `core_vault.
   * self_party_id`. The field keeps its name because every caller in the
   * monorepo reads it as "whose vault is this", which is still true.
   */
  ownerPartyId: string;
  deviceId: string;
  /** The first device's key — the owner's credential. */
  deviceKey: string;
  concepts: Record<string, string>;
}

interface SeedConcept {
  scheme: string;
  notation: string;
  label: string;
}

const SEED_SCHEMES: Record<string, { uri: string; title: string }> = {
  relations: { uri: "urn:duaility:relations", title: "Link relation types" },
  "activity-kinds": {
    uri: "urn:duaility:activity-kinds",
    title: "Activity kinds",
  },
  "spend-categories": {
    uri: "urn:duaility:spend-categories",
    title: "Spend categories",
  },
  flags: { uri: "urn:duaility:flags", title: "Agent flags" },
  // Machine-tag vocabularies (#299): concepts arrive on demand from the
  // enrichment publishers; pre-v10 vaults use the guarded v10 backfill.
  vision: { uri: "urn:centraid:vision", title: "Vision tags (machine)" },
  doctype: { uri: "urn:centraid:doctype", title: "Document types (machine)" },
};
const SEED_CONCEPTS: SeedConcept[] = [
  { scheme: "relations", notation: "same-as", label: "Same as" },
  { scheme: "relations", notation: "about", label: "About" },
  { scheme: "relations", notation: "works-for", label: "Works for" },
  { scheme: "relations", notation: "duplicate-of", label: "Duplicate of" },
  // Cross-referencing relations (#272), also seeded by migration: keep in
  // step.
  { scheme: "relations", notation: "references", label: "References" },
  { scheme: "relations", notation: "attachment-of", label: "Attachment of" },
  // Version lineage (#352): newer content revises older, asserted by the
  // document and note edit commands.
  { scheme: "relations", notation: "revises", label: "Revises" },
  { scheme: "activity-kinds", notation: "meeting", label: "Meeting" },
  { scheme: "activity-kinds", notation: "run", label: "Run" },
  { scheme: "activity-kinds", notation: "sleep", label: "Sleep" },
  { scheme: "activity-kinds", notation: "work", label: "Work session" },
  { scheme: "spend-categories", notation: "groceries", label: "Groceries" },
  { scheme: "spend-categories", notation: "dining", label: "Dining out" },
  { scheme: "spend-categories", notation: "transport", label: "Transport" },
  { scheme: "spend-categories", notation: "gifts", label: "Gifts" },
  { scheme: "flags", notation: "anomaly", label: "Anomaly" },
];

export interface BootstrapVaultOptions {
  ownerName: string;
  baseCurrency?: string;
  deviceName?: string;
  /** Pre-minted id: multi-vault hosts name each vault's directory after it. */
  vaultId?: string;
  vaultName?: string;
  defaultTz?: string;
}

export function bootstrapVault(
  db: VaultDb,
  options: BootstrapVaultOptions
): BootstrapResult {
  const now = nowIso();
  const concepts: Record<string, string> = {};
  const schemeIds: Record<string, string> = {};
  for (const [key, scheme] of Object.entries(SEED_SCHEMES)) {
    const schemeId = uuidv7();
    schemeIds[key] = schemeId;
    db.vault
      .prepare(
        `INSERT INTO core_concept_scheme (scheme_id, uri, title, publisher, version) VALUES (?, ?, ?, 'duaility', '1')`
      )
      .run(schemeId, scheme.uri, scheme.title);
  }
  for (const seed of SEED_CONCEPTS) {
    const conceptId = uuidv7();
    concepts[seed.notation] = conceptId;
    db.vault
      .prepare(
        `INSERT INTO core_concept (concept_id, scheme_id, notation, pref_label, alt_labels_json, broader_concept_id, definition)
         VALUES (?, ?, ?, ?, NULL, NULL, NULL)`
      )
      .run(conceptId, schemeIds[seed.scheme] ?? "", seed.notation, seed.label);
  }
  const ownerPartyId = uuidv7();
  db.vault
    .prepare(
      `INSERT INTO core_party (party_id, kind, display_name, sort_name, birth_date, avatar_content_id, created_at, updated_at)
       VALUES (?, 'person', ?, NULL, NULL, NULL, ?, ?)`
    )
    .run(ownerPartyId, options.ownerName, now, now);
  const vaultId = options.vaultId ?? uuidv7();
  const displayName = options.vaultName ?? `${options.ownerName}'s vault`;
  db.vault
    .prepare(
      `INSERT INTO core_vault (vault_id, self_party_id, display_name, status, base_currency, settings_json, created_at)
       VALUES (?, ?, ?, 'active', ?, '{}', ?)`
    )
    .run(
      vaultId,
      ownerPartyId,
      displayName,
      // No caller passes baseCurrency today, so this USD default IS the
      // currency every real vault displays; readers fall back to USD anyway.
      options.baseCurrency ?? "USD",
      now
    );
  // Enrichment-policy mirror (#352): `gateway` is the default on both domains,
  // same as the settings bag this table shadows. A third-party provider seeing
  // bytes stays gated separately per call and capability (#567).
  for (const domain of ["photos", "docs"] as const) {
    db.vault
      .prepare(
        `INSERT INTO enrich_policy (domain, tier, updated_at) VALUES (?, 'gateway', ?)`
      )
      .run(domain, now);
  }
  // Events require a calendar (schedule.propose_event precondition) but no
  // command mints one — seed a private "Personal" calendar so schedule works
  // from first boot.
  db.vault
    .prepare(
      `INSERT INTO schedule_calendar (calendar_id, owner_party_id, name, color, default_tz, visibility, external_uri)
       VALUES (?, ?, 'Personal', NULL, ?, 'private', NULL)`
    )
    .run(uuidv7(), ownerPartyId, options.defaultTz ?? "UTC");
  const device = enrollDevice(
    db,
    ownerPartyId,
    options.deviceName ?? "first device"
  );
  return {
    vaultId,
    displayName,
    ownerPartyId,
    deviceId: device.deviceId,
    deviceKey: device.deviceKey,
    concepts,
  };
}

export function enrollDevice(
  db: VaultDb,
  ownerPartyId: string,
  name: string,
  trust: "full" | "readonly" = "full"
): { deviceId: string; deviceKey: string } {
  const deviceId = uuidv7();
  const deviceKey = randomBytes(32).toString("hex");
  const now = nowIso();
  // Identity here, authority next door: `access_device` says who the device
  // is, `share_authority` what the member let it do (#883).
  db.vault
    .prepare(
      `INSERT INTO access_device (device_id, owner_party_id, name, platform, public_key, enrolled_at, last_seen_at, sync_cursor)
       VALUES (?, ?, ?, NULL, ?, ?, NULL, NULL)`
    )
    .run(deviceId, ownerPartyId, name, deviceKey, now);
  setDeviceTrust(db.vault, { deviceId, ownerPartyId, trust, now });
  return { deviceId, deviceKey };
}

export function enrollApp(
  db: VaultDb,
  options: {
    name: string;
    riskCeiling?: Risk;
    displayName?: string;
  }
): { appId: string; signingKey: string } {
  const appId = uuidv7();
  const signingKey = randomBytes(32).toString("hex");
  db.vault
    .prepare(
      // `origin` is a one-value vocabulary (#916, ruling ONT-07): an app
      // reaches a vault by being installed, and has had no other door
      // since #799.
      `INSERT INTO access_app (app_id, name, display_name, publisher, manifest_uri, signing_key, status, origin, risk_ceiling, installed_at)
       VALUES (?, ?, ?, NULL, NULL, ?, 'active', 'installed', ?, ?)`
    )
    .run(
      appId,
      options.name,
      options.displayName ?? null,
      signingKey,
      options.riskCeiling ?? "low",
      nowIso()
    );
  return { appId, signingKey };
}

export function enrollAgent(
  db: VaultDb,
  options: {
    name: string;
    modelRef: string;
    version?: string;
    displayName?: string;
  }
): { agentId: string; partyId: string } {
  const now = nowIso();
  const partyId = uuidv7();
  db.vault
    .prepare(
      `INSERT INTO core_party (party_id, kind, display_name, sort_name, birth_date, avatar_content_id, created_at, updated_at)
       VALUES (?, 'agent', ?, NULL, NULL, NULL, ?, ?)`
    )
    .run(partyId, options.displayName ?? options.name, now, now);
  const agentId = uuidv7();
  db.vault
    .prepare(
      `INSERT INTO access_agent (agent_id, party_id, enrollment_key, model_ref, version, enrolled_at, status)
       VALUES (?, ?, ?, ?, ?, ?, 'active')`
    )
    .run(
      agentId,
      partyId,
      options.name,
      options.modelRef,
      options.version ?? "0",
      now
    );
  return { agentId, partyId };
}

export interface ScopeSpec {
  schema: string;
  table?: string;
  verbs: "read" | "read+act" | "act" | "reveal";
  rowFilter?: FilterClause[];
  fieldMask?: string[];
}

File: crates/vault/src/bootstrap.rs
//! Founding a vault's own identity: the vault row and its owner.
//!
//! A freshly created file has a schema and no facts. This writes the two rows
//! nothing else can be written without — `core_vault` and the owner's
//! `core_party` — inside the commit guard, so they are the file's first log
//! commit and a seat bootstrapped from a snapshot of it sees them.
//!
//! It is here and not in `Vault::create` on purpose: creating a file and
//! deciding whose it is are two acts, and `centraid recover` (lane R) restores
//! a file that already has an owner.

use crate::error::Result;
use crate::file::Vault;

/// THE LINK RELATIONS, SEEDED AT FOUND TIME (#272; #1020, D-1020-N7).
///
/// **Relations are VOCABULARY, not caller text**: `core.link_entities` refuses a
/// notation that is not already a concept in the relations scheme, which means
/// the scheme has to exist before the first link — v0 seeds it here
/// (`packages/vault/src/bootstrap.ts:33`-`:64`) and a create-on-demand path
/// would turn "never caller-invented" into "invented on first use".
///
/// **`revises` IS DELIBERATELY ABSENT** (#996 R20(a)). Version lineage was a
/// content→content link asserted by the document and note edit commands — a
/// SECOND history mechanism beside `core_entity_revision`, which [#916] ruled
/// the only one. A version is an occurrence now, the concept that named the edge
/// has no writer, and seeding it would be dormant DDL (ONT-06).
///
/// **The other five seed schemes are still absent** — `activity-kinds`,
/// `spend-categories`, `flags`, `vision` and `doctype`. Docs' folders and flags
/// schemes are created on first use by `crates/vault/src/commands/core.rs`
/// instead, which is a divergence from v0 this lane files as a finding rather
/// than fixes in another slot's schema.
const SEED_RELATIONS: &[(&str, &str)] = &[
    ("same-as", "Same as"),
    ("about", "About"),
    ("works-for", "Works for"),
    ("duplicate-of", "Duplicate of"),
    // Cross-referencing relations (#272), which is what a `[[wikilink]]`
    // compiles to.
    ("references", "References"),
    ("attachment-of", "Attachment of"),
    // THE TWO ANSWERS TO A CROSS-SOURCE MATCH (#996 R20(c) / OQ-12). `same-as`
    // is the acceptance; `distinct-from` is the refusal, and it has to be a
    // relation rather than a dismissed notification because a refusal that is
    // not written down is a proposal the member is shown again tomorrow.
    ("distinct-from", "Distinct from"),
];

/// Seed the relations scheme and its notations. Idempotent over an existing
/// scheme, so re-founding a restored file adds nothing.
fn seed_relation_vocabulary(
    connection: &rusqlite::Connection,
    ids: &dyn crate::clock::Ids,
    now: &str,
) -> Result<()> {
    let uri = crate::commands::core_links::RELATIONS_SCHEME_URI;
    let scheme_id = match connection.query_row(
        "SELECT scheme_id FROM core_concept_scheme WHERE uri = ?1",
        [uri],
        |row| row.get::<_, String>(0),
    ) {
        Ok(scheme_id) => scheme_id,
        Err(_) => {
            let scheme_id = ids.next();
            connection.execute(
                "INSERT INTO core_concept_scheme
                   (scheme_id, uri, title, publisher, version, created_at)
                 VALUES (?1, ?2, 'Link relation types', NULL, '1', ?3)",
                rusqlite::params![scheme_id, uri, now],
            )?;
            scheme_id
        }
    };
    for (notation, label) in SEED_RELATIONS {
        connection.execute(
            "INSERT INTO core_concept
               (concept_id, scheme_id, notation, pref_label, created_at, updated_at)
             VALUES (?1, ?2, ?3, ?4, ?5, ?5)
             ON CONFLICT (scheme_id, notation) DO NOTHING",
            rusqlite::params![ids.next(), scheme_id, notation, label, now],
        )?;
    }
    Ok(())
}

/// What founding produced.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Founded {
    pub vault_id: String,
    pub owner_party_id: String,
}

impl Vault {
    /// Write the vault row and its owner party.
    pub fn found(&self, display_name: &str, owner_name: &str) -> Result<Founded> {
        let now = self.clock().now_text();
        let vault_id = self.ids().next();
        let owner_party_id = self.ids().next();
        self.commit(|tx| {
            tx.set_producer("vault.found");
            // The owner first: `core_vault.self_party_id` points at it, and a
            // vault row with a dangling owner is a file that cannot answer
            // "whose is this".
            tx.connection().execute(
                "INSERT INTO core_party
                   (party_id, kind, display_name, created_at, updated_at)
                 VALUES (?1, 'person', ?2, ?3, ?3)",
                rusqlite::params![owner_party_id, owner_name, now],
            )?;
            tx.connection().execute(
                "INSERT INTO core_vault
                   (vault_id, self_party_id, display_name, status, base_currency,
                    settings_json, created_at, updated_at)
                 VALUES (?1, ?2, ?3, 'active', 'USD', '{}', ?4, ?4)",
                rusqlite::params![vault_id, owner_party_id, display_name, now],
            )?;
            seed_relation_vocabulary(tx.connection(), self.ids(), &now)?;
            Ok(())
        })?;
        Ok(Founded {
            vault_id,
            owner_party_id,
        })
    }

    /// The vault's own id, if it has been founded.
    pub fn vault_id(&self) -> Result<Option<String>> {
        self.read(|connection| {
            Ok(connection
                .query_row(
                    "SELECT vault_id FROM core_vault ORDER BY vault_id LIMIT 1",
                    [],
                    |row| row.get(0),
                )
                .ok())
        })
    }
}
